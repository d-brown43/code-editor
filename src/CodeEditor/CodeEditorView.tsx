import * as React from "react";
import styles from './CodeEditor.module.scss';
import useTextAreaFocus from './useFocus';
import LineNumbers from "./LineNumbers";
import useSyntaxErrorPosition from "./useSyntaxErrorPosition";

type CodeEditorView = {
    program: string;
    setProgram: (program: string, callback?: (program: string) => void) => void;
    programError: ProgramError;
}

export default React.memo(({program, setProgram, programError}: CodeEditorView) => {
    const textArea = React.useRef<HTMLTextAreaElement>(null);
    const [error, setError] = React.useState<ErrorState>({
        x: 0,
        y: 0,
        error: null
    });

    useTextAreaFocus(textArea);
    useSyntaxErrorPosition(programError, textArea, program, setError);

    return (
        <div className={styles.container}>
            <div className={styles.content}>
                {(error.error === null) ? null : (
                    <span
                        style={{
                            position: 'absolute',
                            top: `calc(${error.y}px + 1em)`,
                            left: error.x,
                            color: 'red'
                        }}
                    >
                        <i className="fas fa-chevron-up" />
                        <span style={{marginLeft: 8}}>
                            {error.error}
                        </span>
                    </span>
                )}
                <LineNumbers program={program} />
                <textarea
                    ref={textArea}
                    className={styles.textEditor}
                    value={program}
                    onChange={(e) => setProgram(e.target.value)}
                />
            </div>
        </div>
    )
});
