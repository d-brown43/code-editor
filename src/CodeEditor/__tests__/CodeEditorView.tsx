import * as React from "react";
import CodeEditorView from "../CodeEditorView";
import {shallow} from 'enzyme';

it('matches snapshot', () => {
    const program = `console.log('Hello World');`;
    const setProgram = jest.fn();

    const codeEditorView = shallow(
        <CodeEditorView
            programError={false}
            program={program}
            setProgram={setProgram}
        />
    );
    expect(codeEditorView).toMatchSnapshot();
});

it('matches snapshot', () => {
    const program = `console.log('Hello World');`;
    const setProgram = jest.fn();

    const codeEditorView = shallow(
        <CodeEditorView
            programError={{
                location: {
                    line: 10,
                    column: 1
                },
                message: 'Test Message'
            }}
            program={program}
            setProgram={setProgram}
        />
    );
    expect(codeEditorView).toMatchSnapshot();
});

it(`sets state to render a warning when there's a syntax error`, () => {

});
