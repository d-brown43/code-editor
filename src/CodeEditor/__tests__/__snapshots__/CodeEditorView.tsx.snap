// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="content"
  >
    <LineNumbers
      program="console.log('Hello World');"
    />
    <textarea
      className="textEditor"
      onChange={[Function]}
      value="console.log('Hello World');"
    />
  </div>
</div>
`;

exports[`matches snapshot 2`] = `
<div
  className="container"
>
  <div
    className="content"
  >
    <LineNumbers
      program="console.log('Hello World');"
    />
    <textarea
      className="textEditor"
      onChange={[Function]}
      value="console.log('Hello World');"
    />
  </div>
</div>
`;
